{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyP35/YeTjwCKefrYFy9SohO",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/dramitprabhumd-cpu/60-Days-of-Biopython/blob/main/Phase_1/Day02_Central_Dogma.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jQqbgdU03uzI",
        "outputId": "66044738-5d2b-4179-8b69-53421f180d9f"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Using cached biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.12/dist-packages (from biopython) (2.0.2)\n",
            "Downloading biopython-1.86-cp312-cp312-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl (3.2 MB)\n",
            "\u001b[?25l   \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m0.0/3.2 MB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\r\u001b[2K   \u001b[91mâ”â”\u001b[0m\u001b[90mâ•º\u001b[0m\u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m0.2/3.2 MB\u001b[0m \u001b[31m5.3 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K   \u001b[91mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m\u001b[91mâ•¸\u001b[0m\u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m1.3/3.2 MB\u001b[0m \u001b[31m19.2 MB/s\u001b[0m eta \u001b[36m0:00:01\u001b[0m\r\u001b[2K   \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m27.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.86\n"
          ]
        }
      ],
      "source": [
        "!pip install biopython"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# ğŸ§¬ Biopython 60-Day Portfolio: Day 2\n",
        "## Project: The Central Dogma (Transcription & Translation)\n",
        "\n",
        "### ğŸ¯ Objective\n",
        "Automate the biological process of converting Genetic information into functional Proteins using Biopython's built-in methods.\n",
        "\n",
        "### ğŸ› ï¸ Key Concepts\n",
        "* **Transcription:** Converting the \"Coding Strand\" of DNA into mRNA using `.transcribe()`.\n",
        "* **Translation:** Converting mRNA (or DNA) into an Amino Acid sequence using `.translate()`.\n",
        "* **Stop Codons:** Identifying the end of a protein sequence (represented as `*`).\n",
        "\n",
        "---\n",
        "\n",
        "### ğŸ““ Notes\n",
        "* Biopython follows the **Standard Genetic Code** by default.\n",
        "* If you are working with Mitochondrial DNA, you must specify the table: `dna_seq.translate(table=\"Vertebrate Mitochondrial\")`."
      ],
      "metadata": {
        "id": "beV_bVhs4GQf"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.Seq import Seq"
      ],
      "metadata": {
        "id": "O0CBX5374hg7"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# 1. Starting with coding DNA sequence\n",
        "dnaSeq = Seq('ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG')\n",
        "print (f'DNA: {dnaSeq}')\n",
        "\n",
        "# 2. DNA to mRNA (Transcription)\n",
        "# In Biopython, this replaces Thymine (T) with Uracil (U)\n",
        "mRNA = dnaSeq.transcribe()\n",
        "print(f'mRNA: {mRNA}')\n",
        "\n",
        "# 3. mRNA to Protein Seq (Tranlation)\n",
        "# The stop codon (*) is usually represented by an asterisk\n",
        "proteinSeq = mRNA.translate()\n",
        "print(f'Protein: {proteinSeq}')\n",
        "\n",
        "# 4. Biopython allows direct translation from a DNA sequence\n",
        "proteinSeqDirect = dnaSeq.translate()\n",
        "print(f'Protein (Direct from DNA): {proteinSeqDirect}')\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "TKm5WBOg4pmw",
        "outputId": "09add431-31cc-40af-ad32-6a84e0ae6828"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "DNA: ATGGCCATTGTAATGGGCCGCTGAAAGGGTGCCCGATAG\n",
            "mRNA: AUGGCCAUUGUAAUGGGCCGCUGAAAGGGUGCCCGAUAG\n",
            "Protein: MAIVMGR*KGAR*\n",
            "Protein (Direct from DNA): MAIVMGR*KGAR*\n"
          ]
        }
      ]
    }
  ]
}